### ISSUE TITLE
💡 [REQUEST] - Parameter cutsomization for FASTP

### ISSUE THREAD

**Issue by gregdenay** [bio-raum/FooDMe2#88]
### Is your feature request related to a problem?

There is currently no way to change the FASTP parameters and QC trimmming is very basic.
Currently QC trimming is only removing reads with N and shorter than 40bp. This is hardcoded.

### Describe the solution you'd like

- We should remove the `-n` parameter since Ns and EE filters are already enforced in either of the DADA or VSEARCH workflows.
- Expose the `-l` parameter for user input
- Add `-3 --cut_tail_window_size XX --cut_tail_mean_quality XX` toadd a sliding window QC trimming on the 3' side of reads 
- Expose paremeters for `cut_tail_window_size` and  `cut_tail_mean_quality` for user input
- Add a parameter allowing to pass more command line arguments to fastp (like for cutadapt)?

**Comment by gregdenay** [comment]
Should be careful how this interacts with cutadapt though, maybe we need to switch up the order of the processes (Cutapdapt > Fastp instead of first fastp)

**Comment by marchoeppner** [comment]
Right, yes, this is a potential issue of course. FastPs main use at the moment are a) gathering metrics for the now defunct MultiQC report and b) removing sequencing adapters (which can sometime remain in the reads at low frequency). Any other trimming, as you say, is super basic and delegated to Cutadapt. 

Exposing additional parameters is of course an option, although it would interfere with the downstream processing of the reads. 

If we decide to expose options (again, I am not sure this is useful), then in the form of 

`--fastp_cut_tail_mean_quality` 

and similar.

**Comment by marchoeppner** [comment]
My guess is that FastP would be the better choice to perform most of that - partly because it is a lot faster than Cutadapt. Cutadapt at the moment is also a pretty crucial building block of the pipeline, maybe best to have it do as few things as possible to not make it more complicated. The only thing FastP doesn't have, I think, is the ee-filtering (it has an -e option, which sounds somewhat simlar but probably uses a slightly different calculation). 

Whether we can move some of the filtering currently done in vsearch and dada to cutadapt (or fastP) - not sure if this would interfere with the algorithms? I am guessing the actual implementation will be slightly different and at least for Dada2 it sounded like there are some assumptions being made internally that would benefit from Dada2 doing as many of the processing steps as possible (I could be wrong).

**Comment by marchoeppner** [comment]
Quick question - would you prefer that all the above options are exposed individually, or could we simply to 
```
--fastp_options somethinghere
```

so users can manipulate fastp fully without being limited to a small number of parameters? We could set "reasonable" defaults to start with. 

Advantage of a generic cli option is, as mentioned, the possibility to change all aspects of fastp, but the downside is that all defaults get overwritten as soon as the option is used.

**Comment by gregdenay** [comment]
I think that is fine. Documentation just has to be clear on this. 
It is more of an advanced option anyways so users should know what they do.

**Comment by gregdenay** [comment]
```
--cut_tail_window_size 4
--cut_tail_mean_quality 25
```

had no adverse effect in FooDMe1, i guess the min quality could even be raised to 30. Worse that can happen is that the reads get very short (which would be a sign of something going very wrong during sequencing since the reads are trimmed from 3' to 5' with this option). So we should add  `--length_required` (I assume 50 should be enough for the meat method).

The --cut_front and --cut_right methods are pretty aggressive, cut_tail is the one we shoudl use. We can safely ignore the deduplication warning here since fwd and rev reads are first merged. The only problem is if the trimmed reads are too short to merge, but then again, that would be the sign that something went wrong during sequencing. 

If this makes difficulties for other barcodes we can always lower the trheshold or bypass this step altogether.

**Comment by marchoeppner** [comment]
Added in [c0b99e8](https://github.com/bio-raum/FooDMe2/commit/c0b99e820b2b576c1a406afc825b734e83abe548)

But - setting the options as suggested deteriorates the sample metrics quite a bit.

Old defaults:
![grafik](https://github.com/user-attachments/assets/fe50804a-e414-4ba9-9f11-911f0e1bd795)

New settings:
![grafik](https://github.com/user-attachments/assets/96803fe4-f471-4ba6-8246-f99938209214)

I have introduced the option --fastp_options, but have left the settings to fastp as before. Maybe this needs some more investigating.

**Comment by gregdenay** [comment]
Yes we need to switch the order of fastp and cutadapt so that primer sequences are trimmed **before** any kind of quality trimming, otherwise the primer finding rate drops.

**Comment by gregdenay** [comment]
I saw that we had Fastp first in order to check the insert size distribution before primer trimming in order to produce some warnings should the insert size be larger than read length.  Since we only warn and not stop anything we could still do that after primer trimming. 
Alternatively I would suggest leaving the first Fastp step without anyfiltering and addign a second one after cutadapt with the filters above. We could switch the first fastp step with something faster but it seems that fastp is the fastest solution?

**Comment by marchoeppner** [comment]
Hi,

yes, moving fastp from its current place without replacement would be an issue (also: trimming of illumina sequencing adapters). 

So having a second instance downstream of cutadapt is probably the only solution. Or have cutadapt to the quality trimming after all (which might be a pain to do because of the convoluted structure we need for the primer trimming). 

To my knowledge there is no faster solution than fastP; although metabarcoding data is so small in size that it probably makes no significant difference. For human WGS data, the difference is substantial though (again, not a concern here).

**Comment by gregdenay** [comment]
[e87096d](https://github.com/bio-raum/FooDMe2/commit/e87096db7889d743b19ac1da9af2df7b0407c55d)

adds an additonal FASTZP step after cutadapt that uses the options given by `--fastp_options`, the first fastp rule only does adpater detection and outputs metrics.

The change has no effect on 16S meat data (as expected).

Still need to change the reporting to read in the trimming metrics from the second fastp step.

**Comment by marchoeppner** [comment]
Great!

I took the liberty of adjusting those changes a bit. There isn't technically a reason to have two separate fastp modules, as far as I can see. So I reverted that back to one module, as before, and added an additional external parameter "suffix", which gets set for the first "metric" fastp stage (although there isn't an actual risk of having name collisions, I think). 

Bascially, you can import any module multiple times under different names; and those names can then also be used in the config file to sent different parameters to that "version" of the module.

Have a look and let me know if that makes sense. 
[20ea597](https://github.com/bio-raum/FooDMe2/commit/20ea597c433905a062cbb9f5cd21aaebcedb8f79)


/M

**Comment by gregdenay** [comment]
Smart :) 

The report now take the fastp metrics before and after the whole primer trimming + quality trimming steps for the calculation of the filterdd reads and the read quality plots. I also added a tabset for the insert size histogram to see the amplicon size before and after primer trimming.
